### PR TITLE
Auditory First Version

### PR DESCRIPTION
**Pull request by yaniaiciin2**
ADD:
-> Now it creates auditory transaction registry when a event is generated
 -> Added JPA support for transaction DB
-> Created "Transaction" class
CHANGE:
-> Fixed some reactive controllers
-> Changed application.yml configuration for audit DB

### CODE REVIEW

File: src/main/java/es/in2/blockchain/connector/core/configuration/JpaConfig.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter;
+
+import javax.sql.DataSource;
+
+@Configuration
+@EnableJpaAuditing
+public class JpaConfig {
+
+    @Bean
+    public LocalContainerEntityManagerFactoryBean entityManagerFactory(DataSource dataSource) {
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/configuration/JpaConfig.java:18]: @yaniaiciin2  'em' is not a good definition if we follow Clean Code.

File: src/main/java/es/in2/blockchain/connector/core/configuration/JpaConfig.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        em.setPackagesToScan("es.in2.blockchain.connector.core.domain");
+
+        HibernateJpaVendorAdapter vendorAdapter = new HibernateJpaVendorAdapter();
+        em.setJpaVendorAdapter(vendorAdapter);
+
+        return em;
+    }
+
+    @Bean
+    public JpaTransactionManager transactionManager(EntityManagerFactory entityManagerFactory) {
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/configuration/JpaConfig.java:30]: Same as above, JpaTrasnsactionManager jpaTransactionManager ;)

File: src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+@Getter
+@Setter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+@Audited
+public class Transaction {
+
+    @Id
+    @GeneratedValue
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java:23]: Use strategy to add @GeneratedValue(strategy = GenerationType.UUID)

File: src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    @Column(columnDefinition = "BINARY(16)")
+    private UUID id;
+
+    @Audited
+    private String entityId;
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java:28]: This entityID is just the ID or all the location URL?

File: src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Audited
+    private String status;
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java:31]: Better if Status is an ENUM that brings distinct defined status types (RECEIVED, SENT, POST TO BLOCKCHAIN, ERROR, etc.)

File: src/main/java/es/in2/blockchain/connector/core/service/impl/TransactionServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.NoSuchElementException;
+import java.util.UUID;
+
+@Service
+public class TransactionServiceImpl implements TransactionService {
+    private final TransactionRepository transactionRepository;
+    @Autowired
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/service/impl/TransactionServiceImpl.java:15]: @Autowired is forbidden. We use injection through constructor. Use the annotation @RequiredArgsConstructor

**Review by oriolcanadesin2 — CHANGES_REQUESTED** [review]

File: src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java (hunk 4 above)
Comment (yaniaiciin2) [reply src/main/java/es/in2/blockchain/connector/core/domain/Transaction.java:28]: Just the ID

File: src/main/java/es/in2/blockchain/connector/core/utils/AuditStatus.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
+public enum AuditStatus {
+    RECEIVED("RECIEVED"),
+    POST_TO_BLOCKCHAIN("POST TO BLOCKCHAIN"),
+    ERROR("ERROR"),
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/utils/AuditStatus.java:11]: Delete ERROR status

File: src/main/java/es/in2/blockchain/connector/core/service/impl/TransactionServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+                .entityId(entityId)
+                .entityHash(entityHash)
+                .dataLocation(datalocation)
+                .status(AuditStatus.RECEIVED.getDescription())
+                .build();
+
+        return  transactionRepository.save(transaction);
+    }
+
+    @Override
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/service/impl/TransactionServiceImpl.java:30]: Old:
	
	@Override
    public Transaction editTransactionAttribute(UUID id, String newAttributeValue) {

        Transaction existingTransaction = transactionRepository.findById(id);

        if (existingTransaction != null) {
            existingTransaction.setStatus(newAttributeValue);
            return transactionRepository.save(existingTransaction);
        }

        throw new NoSuchElementException("Transaction not found.");

    }
	
	New:
	
	@Override
    public Transaction editTransactionAttribute(UUID id, String newAttributeValue) {
        Transaction transactionFound = transactionRepository.findById(id);
		checkIfTransactionExist(transactionFound);
        existingTransaction.setStatus(newAttributeValue);
        return transactionRepository.save(existingTransaction);
    }
	
	private void checkIfTransactionExist(transactionFound Transaction) {
		// TODO: check if is null or empty
		if (transactionFound == null) { 
			throw new NoSuchElementException("Transaction not found.");
        }
	}

File: src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java [hunk 9]
```diff
@@ -28,18 +31,29 @@ public class OnChainEntityServiceImpl implements OnChainEntityService {
     private final HashLinkService hashLinkService;
     private final BlockchainNodeIConfig blockchainNodeIConfig;
     private final BlockchainNodeProperties blockchainNodeProperties;
+    private final TransactionService transactionService;
 
     @Override
     public void createAndPublishEntityToOnChain(OrionLdNotificationDTO orionLdNotificationDTO) {
         // Create On-Chain Entity DTO
         DomeEvent domeEvent;
+        Transaction transaction;
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java:40]: Received: datalocation

File: src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java [hunk 10]
```diff
@@ -28,18 +31,29 @@ public class OnChainEntityServiceImpl implements OnChainEntityService {
...
         try {
             domeEvent = createOnChainEntityDTO(orionLdNotificationDTO);
+            transaction = transactionService.createTransaction(orionLdNotificationDTO.getId(), hashLinkService.extractHashLink(domeEvent.getDataLocation()), domeEvent.getDataLocation());
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java:43]: Created + Hash

File: src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java [hunk 11]
```diff
@@ -28,18 +31,29 @@ public class OnChainEntityServiceImpl implements OnChainEntityService {
...
         } catch (JsonProcessingException e) {
             throw new RequestErrorException("Error creating On-Chain Entity DTO: " + e.getMessage());
         }
         // Publish On-Chain Entity DTO to Blockchain Node Interface
-        publishDomeEvent(domeEvent);
+        transactionService.editTransactionAttribute(transaction.getId(), AuditStatus.CREATED.getDescription());
+        try{
+            publishDomeEvent(domeEvent);
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java:50]: Published

**Review by oriolcanadesin2 — CHANGES_REQUESTED** [review]

File: src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java [hunk 12]
```diff
@@ -28,18 +32,25 @@ public class OnChainEntityServiceImpl implements OnChainEntityService {
...
 
     @Override
     public void createAndPublishEntityToOnChain(OrionLdNotificationDTO orionLdNotificationDTO) {
         // Create On-Chain Entity DTO
         DomeEvent domeEvent;
+        Transaction transaction;
+        transaction = transactionService.createTransaction(orionLdNotificationDTO.getId(), " ", hashLinkService.createHashLink(orionLdNotificationDTO.getId(),parseNotificationData(orionLdNotificationDTO)));
         try {
             domeEvent = createOnChainEntityDTO(orionLdNotificationDTO);
+            transactionService.editTransactionHash(transaction.getId(), domeEvent.getDataLocation());
```
Comment (oriolcanadesin2) [inline src/main/java/es/in2/blockchain/connector/core/service/impl/OnChainEntityServiceImpl.java:45]: Why we have 2 methods to edit a Transaction? Is it possible to group the methods into a single one, editTransaction()?

**Review by oriolcanadesin2 — CHANGES_REQUESTED** [review]

**Pull request merged**